### PR TITLE
Refactoring and Improvements to internals of bounding boxes

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

When writing ASDF converters for bounding box and compound bounding box in #12696. I found that in order to properly support converters, these classes should be able to initialize with data from an asdf file without a reference to their model. This PR makes this change possible.

This PR also cleans up the how `ignored` inputs are handled for `fix_inputs` and `CompoundBoundingBox`. Additionally, it switches all the input key storage from using integers to using the string names tracked in the models which simplifies direct examination of the objects without any changes to the external interfaces.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/modeling/bounding_box.py [hunk 1]
```diff
@@ -578,47 +632,116 @@ class ModelBoundingBox(_BoundingDomain):
...
+    def _verify_intervals_dict(self, _external_ignored: List[str] = None):
+        if _external_ignored is None:
+            _external_ignored = []
+
+        intervals = self._pop_intervals()
+
+        for key, value in intervals.items():
+            name = self._get_name(key)
+            if name in _external_ignored:
+                raise ValueError(f"Interval: {name} is being externally ignored!")
```
Comment (eslavich) [inline astropy/modeling/bounding_box.py:661]: Will it cause trouble that we leave self._intervals empty after hitting this error?  Ordinarily I'd expect that a validation failure would reject the change and leave the object as it was.

File: astropy/modeling/bounding_box.py [hunk 2]
```diff
@@ -578,47 +632,116 @@ class ModelBoundingBox(_BoundingDomain):
...
+
+            self[name] = value
+
+        if len(self.intervals) != 0 and len(self.intervals) != len(names := self._interval_names(_external_ignored)):
```
Comment (eslavich) [inline astropy/modeling/bounding_box.py:665]: This is the first walrus operator I've seen in the wild!

**Review by eslavich — COMMENTED** [review]
I don't know enough about bounding boxes to follow this, but I thought of a question to ask...

File: astropy/modeling/bounding_box.py (hunk 1 above)
Comment (WilliamJamieson) [reply astropy/modeling/bounding_box.py:659]: I believe this is addressed by 0dfa0c3

**Review by eslavich — COMMENTED** [review]
I worked through some of bounding_box.py, but I want to see what you think of these suggestions before continuing.

File: astropy/modeling/bounding_box.py [hunk 3]
```diff
@@ -188,21 +189,39 @@ class _BoundingDomain(abc.ABC):
         on the inputs and returns a complete output.
     """
 
-    def __init__(self, model, ignored: List[int] = None, order: str = 'C'):
-        self._model = model
-        self._ignored = self._validate_ignored(ignored)
+    def __init__(self, model = None, ignored: List[str] = None, order: str = 'C'):
```
[the remark this answers is not in the document]
Comment (WilliamJamieson) [reply astropy/modeling/bounding_box.py:192]: This would be the simplest solution if we didn't what to make all these changes.

**Pull request closed without merging**